### PR TITLE
Block API: Ensure backwards compatibility for block matchers

### PR DESCRIPTION
**Pull request by youknowriad**
with #2854 it's no longer possible to use attribute matchers to define the block attributes. This PR adds a way to ensure backwards compatibility while warning about the "deprecated" status of these matchers.

**Testing instructions**

 - Update any block's attributes definition to the old matchers signature (example: `source.attr( 'myattribute' )`)
 - Load Gutenberg
 - the block should still work as expected
 - A warning should show up in the console when loading the editor.

### CODE REVIEW

File: blocks/api/source.js [hunk 1]
```diff
@@ -1,38 +1,66 @@
-/**
- * WordPress dependencies
- */
-import { createElement } from '@wordpress/element';
+function warnAboutDeprecatedMatcher() {
+	// eslint-disable-next-line no-console
+	console.warn(
+		'Attributes matchers are deprecated and they will be remove in future version of Gutenberg.' +
```
Comment (aduth) [inline blocks/api/source.js]: Typo: "remove" -> "removed"
Grammar: "in future version" -> "in a future version"

File: blocks/api/source.js [hunk 2]
```diff
@@ -1,38 +1,66 @@
...
+		'Please update your attributes definition https://wordpress.org/gutenberg/handbook/reference/attributes/'
```
Comment (aduth) [inline blocks/api/source.js]: Grammar: There should be a space between these two sentences, either prefixing this string or suffixing the previous line. (or separate arguments to `console.warn` which is maybe not the best behavior to rely on)

File: blocks/library/button/index.js [hunk 3]
```diff
@@ -156,6 +156,7 @@ class ButtonBlock extends Component {
 			</span>,
 			focus && (
 				<form
+					key="form-link"
```
Comment (aduth) [inline blocks/library/button/index.js]: Was this change intentional?

File: blocks/api/registration.js [hunk 4]
```diff
@@ -125,6 +125,17 @@ export function registerBlockType( name, settings ) {
 		...settings,
 	};
 
+	// Resolve deprecated attributes
+	settings.attributes = mapValues( settings.attributes, ( attribute ) => {
+		if ( isFunction( attribute.source ) ) {
```
Comment (aduth) [inline blocks/api/registration.js]: Do we need to return functions from the deprecated matcher? Why not just return the object, and skip this step?

Comment (youknowriad) [reply blocks/api/registration.js]: Because it won't match the current API.

it will produce: `source: { source: 'text' }` which is not the correct attribute definition

File: blocks/library/button/index.js (hunk 3 above)
Comment (youknowriad) [reply blocks/library/button/index.js]: Oops, this was a bug fix I wanted to publish separately and accidentally made it to this PR

Comment (youknowriad) [reply blocks/library/button/index.js]: I can extract it to a separate PR but maybe not worth it

**Review by aduth — APPROVED** [review]

File: blocks/api/registration.js (hunk 4 above)
Comment (aduth) [reply blocks/api/registration.js]: This could also be implemented as a filter hook, which has the added advantage of being simpler to drop 😄

Comment (youknowriad) [reply blocks/api/registration.js]: Great idea

File: blocks/hooks/matchers.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+				...attribute.source(),
+			};
+		}
+		return attribute;
+	} );
+
+	return settings;
+}
+
+export default function anchor( { addFilter } ) {
```
Comment (aduth) [inline blocks/hooks/matchers.js]: Copypasta.
